frontend/agents/gpt-4_python_compute/generate/computations.py

```diff
@@ -58,62 +46,161 @@ def extract_python_code(response):
+    except openai.PermissionDeniedError as e:
+        raise Exception(f"Permission denied: {str(e)}")
+    except openai.NotFoundError as e:
+        raise Exception(f"Resource not found: {str(e)}")
+    except openai.UnprocessableEntityError as e:
+        raise Exception(f"Unprocessable entity: {str(e)}")
+    except openai.InternalServerError as e:
+        raise Exception("OpenAI server error. Please try again later.")
+    except Exception as e:
+        raise Exception(f"Unexpected error: {str(e)}")
 
 if __name__ == "__main__":
     try:
         # Read JSON string from stdin
         input_json = sys.stdin.read()
-        
+
         # Parse the JSON input
         data = json.loads(input_json)
 
         # Extract the arguments from the parsed JSON
         user_prompt = data['userMessage']
-        # model_version = data.get('selectedModel', 'gpt-4')
         conversation_history = data.get('conversationHistory', [])
         apiKey = data["apiKey"]
 
+        # Validate API key
+        if not apiKey or not isinstance(apiKey, str) or len(apiKey.strip()) == 0:
+            raise Exception("Invalid or missing API key")
+
+        # Validate user prompt
+        if not user_prompt or not isinstance(user_prompt, str) or len(user_prompt.strip()) == 0:
+            raise Exception("Invalid or empty user prompt")
+
         # Call the compute function and get the result
-        result = compute(user_prompt, 'gpt-4o', conversation_history, apiKey)
+        result = compute(user_prompt, 'chatgpt-4o-latest', conversation_history, apiKey)
 
         # Print the result as a JSON string
         print(json.dumps(result))
     except Exception as e:
         # Capture and print the full stack trace
         error_traceback = traceback.format_exc()
-        print(json.dumps({"error": str(e), "traceback": error_traceback}))
+
+        # Determine the appropriate status code
+        status_code = 500  # Default to internal server error
+        if isinstance(e, openai.RateLimitError):
+            status_code = 429
+        elif isinstance(e, openai.AuthenticationError):
+            status_code = 401
+        elif isinstance(e, openai.PermissionDeniedError):
+            status_code = 403
+        elif isinstance(e, openai.NotFoundError):
+            status_code = 404
+        elif isinstance(e, openai.BadRequestError):
+            status_code = 400
+        elif isinstance(e, openai.UnprocessableEntityError):
+            status_code = 422
+        elif "Invalid or missing API key" in str(e):
+            status_code = 401
+        elif "Invalid or empty user prompt" in str(e):
+            status_code = 400
+
+        error_response = {
+            "error": {
+                "message": str(e),
+                "status_code": status_code,
+                "traceback": error_traceback,
+                "type": e.__class__.__name__
+            }
+        }
+        print(json.dumps(error_response))
```

Comment:
I think you can print this to `stderr` then `exit(1)` so that `spawnAsync` rejects the promise running this script

Reply:
done